### PR TITLE
Fix ci workflow timeout

### LINKED ISSUE #5389: [maintenance] Adjust test job timeouts in GitHub Actions CI/CD jobs
🐞 **Describe the bug**

The CI doesn't complete currently because the test jobs are capped to run under 15 minutes and they now get killed on timeout. I believe this is called by the new pip resolver doing backtracking which is rather time-consuming.

💡 **To Reproduce**

It's visible in any recent CI build.

💡 **Expected behavior**

The jobs should complete. Carefully bumping the timeout in the workflow file should be enough.

📋 **Logs/tracebacks**

Example: https://github.com/aio-libs/aiohttp/runs/1669008218?check_suite_focus=true

📋 **Your version of the Python**

N/A

📋 **Your version of the aiohttp/yarl/multidict distributions**

repo master

📋 **Additional context**

N/A

### PR DESCRIPTION
**Pull request by greshilov**
## What do these changes do?

Increasing ci workflow timeout.

**UPDATE** Timeout is not changed, but `pip-tools` is used to pin dependencies and solve timeout issue.

## Are there changes in behavior for the user?

No.

## Related issue number

Fixes #5389 

## Checklist

- [x] I think the code is well written
- [x] Unit tests for the changes exist
- [x] Documentation reflects the changes
- [x] If you provide code modification, please add yourself to `CONTRIBUTORS.txt`
  * The format is &lt;Name&gt; &lt;Surname&gt;.
  * Please keep alphabetical order, the file is sorted by names.
- [x] Add a new news fragment into the `CHANGES` folder
  * name it `<issue_id>.<type>` for example (588.bugfix)
  * if you don't have an `issue_id` change it to the pr id after creating the pr
  * ensure type is one of the following:
    * `.feature`: Signifying a new feature.
    * `.bugfix`: Signifying a bug fix.
    * `.doc`: Signifying a documentation improvement.
    * `.removal`: Signifying a deprecation or removal of public API.
    * `.misc`: A ticket has been closed, but it is not of interest to users.
  * Make sure to use full sentences with correct case and punctuation, for example: "Fix issue with non-ascii contents in doctest text files."

---
This change is [Reviewable](https://reviewable.io/reviews/aio-libs/aiohttp/5408)

### CODE REVIEW

**Review by webknjaz — APPROVED** [review]

**Comment by webknjaz** [comment]
@greshilov Do you want to explore using `pip-tools`?

**Comment by greshilov** [comment]
> @greshilov Do you want to explore using `pip-tools`?

Looks like i have no choice. I'll try a few more times with `constraints.txt`, but `pip-tools` will solve this problem for sure.

**Comment by greshilov** [comment]
@webknjaz `pip-tools` works like a charm, thank you for suggestioning this package. Please, take a look.

- `dev.txt` renamed to `dev.in`. `dev.txt` added to `.gitignore` since it autogenerated.
- I've updated `cryptography` version, because `pip-compile` tool failed to resolve dependencies with old one.
- Makefile modified accordingly.

**Comment by webknjaz** [comment]
> * `dev.txt` renamed to `dev.in`. `dev.txt` added to `.gitignore` since it autogenerated.

This is a bad idea. Both files must be committed. Moreover, dependabot automatically recognizes the pairs of `.in` + `.txt` files and autoupdates the latter.

File: .gitignore [hunk 1]
```diff
@@ -63,3 +63,4 @@ sources
 var/*
 venv
 virtualenv.py
+requirements/dev.txt
```
Comment (webknjaz) [inline .gitignore]: ignoring the lockfiles contradicts the idea of having pins for reproducible envs.
```suggestion
```

File: Makefile [hunk 2]
```diff
@@ -61,7 +61,11 @@ aiohttp/%.c: aiohttp/%.pyx $(call to-hash,$(CYS)) aiohttp/_find_header.c
 .PHONY: cythonize
 cythonize: .install-cython $(PYXS:.pyx=.c)
 
-.install-deps: .install-cython $(PYXS:.pyx=.c) $(call to-hash,$(CYS) $(REQS))
+.compile-deps:
+	pip install pip-tools
+	pip-compile -q requirements/dev.in
```
Comment (webknjaz) [inline Makefile]: This shouldn't be done on each install. Each update should be performed in separate tested PRs.

File: requirements/test.txt [hunk 3]
```diff
@@ -2,7 +2,7 @@
 -r base.txt
 Brotli==1.0.9
 coverage==5.3.1
-cryptography==3.2.1; platform_machine!="i686" and python_version<"3.9" # no 32-bit wheels; no python 3.9 wheels yet
+cryptography==3.3.1; platform_machine!="i686" and python_version<"3.9" # no 32-bit wheels; no python 3.9 wheels yet
```
Comment (webknjaz) [inline requirements/test.txt:5]: I don't think this is related to the PR.

**Review by webknjaz — CHANGES_REQUESTED** [review]

File: .pre-commit-config.yaml [hunk 4]
```diff
@@ -29,7 +29,6 @@ repos:
   - id: end-of-file-fixer
     exclude: >-
       ^docs/[^/]*\.svg$
-  - id: requirements-txt-fixer
```
Comment (webknjaz) [inline .pre-commit-config.yaml:32]: Instead of removing the hook, use include/exclude to point it to `.in` files only

File: CHANGES/5389.bugfix [hunk 5]
```diff
@@ -0,0 +1 @@
+Use pip-tools to pin dependencies
```
Comment (webknjaz) [inline CHANGES/5389.bugfix]: It's not a bugfix, aiohttp-wise. It's probably misc or something like that.

File: .gitignore (hunk 1 above)
Comment (greshilov) [reply .gitignore]: Done.

File: Makefile (hunk 2 above)
Comment (greshilov) [reply Makefile]: Done.

File: requirements/test.txt (hunk 3 above)
Comment (greshilov) [reply requirements/test.txt:5]: `pip-compile` is not able to pin dependencies without this change.

File: .pre-commit-config.yaml (hunk 4 above)
Comment (greshilov) [reply .pre-commit-config.yaml:32]: Done

File: CHANGES/5389.bugfix (hunk 5 above)
Comment (greshilov) [reply CHANGES/5389.bugfix]: Done

File: .pre-commit-config.yaml [hunk 6]
```diff
@@ -30,6 +30,8 @@ repos:
     exclude: >-
       ^docs/[^/]*\.svg$
   - id: requirements-txt-fixer
+    exclude: >-
+      ^requirements/dev.txt
```
Comment (webknjaz) [inline .pre-commit-config.yaml:34]:
```suggestion
      ^requirements/dev.txt$
```

File: Makefile [hunk 7]
```diff
@@ -135,9 +135,16 @@ doc:
...
 	@pip install -U 'pip'
+
+.PHONY: compile-deps
+compile-deps: .update-pip
+	@pip install pip-tools
+	@pip-compile --allow-unsafe -q requirements/dev.in
+
+.PHONY: install
+install: .update-pip
 	@pip install -Ur requirements/dev.txt
```
Comment (webknjaz) [inline Makefile:148]: The canonical use of the lockfiles is as follows:
```suggestion
	@pip install -r requirements/dev.in -c requirements/dev.txt
```

**Comment by webknjaz** [comment]
Urgh... auto-merge got triggered when I posted comments :(

**Comment by webknjaz** [comment]
@greshilov mind backporting this?

**Comment by webknjaz** [comment]
Also, it's not clear how well dependabot works with files included via `-r`. It may be necessary to apply pip-tools there as well.

**Comment by greshilov** [comment]
> Urgh... auto-merge got triggered when I posted comments :(

Strange behaviour.

Should I create another PR with hotfixes?

**Comment by webknjaz** [comment]
yes, please

**Pull request merged**